New dio tests segfaulting in Fedora CI

**Describe the bug**
Starting in early April 2026, the Fedora weekly CI run starts failing on the `cmake test` step.

**To Reproduce**
Steps to reproduce the behavior:
- As this time its not been reproduced in a non-CI environment. Suspect Docker image.

**Expected behavior**
The `ctests` we create work as expected.

**Screenshots**
```
Run cmake --build build --target test
[0/1] Running tests...
Test project /__w/freeciv21/freeciv21/build
    Start 1: test_utility_paths
1/4 Test #1: test_utility_paths ...............   Passed    0.01 sec
    Start 2: test_dio
2/4 Test #2: test_dio .........................Subprocess aborted***Exception:   0.15 sec
    Start 3: test_rulesets
3/4 Test #3: test_rulesets ....................   Passed    4.81 sec
    Start 4: test_server_cli
4/4 Test #4: test_server_cli ..................   Passed    1.80 sec

75% tests passed, 1 tests failed out of 4

Total Test time (real) =   6.78 sec

The following tests FAILED:
Errors while running CTest
Output from these tests are in: /__w/freeciv21/freeciv21/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
	  2 - test_dio (Subprocess aborted)
FAILED: [code=8] CMakeFiles/test.util 
cd /__w/freeciv21/freeciv21/build && /usr/bin/ctest --force-new-ctest-process
ninja: build stopped: subcommand failed.
Error: Process completed with exit code 8.
```


**Platform and version (please complete the following information):**
 - OS: Linux
 - Freeciv21 version: `master`
 - Ruleset/Longturn game (if applicable): N/A

**Additional context**
Dio is new as of #2844 and #2845

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -62,10 +62,13 @@ jobs:
           . venv/bin/activate
           pip install -r docs/requirements.txt
       - name: Configure
+        # Set ALWAYS_ROOT because the container user is root and creating an
+        # unpriviledged user would be too much work.
         run: |
           cmake . -B build -G Ninja \
             -DCMAKE_BUILD_TYPE=${{ matrix.build_type }} \
-            -DCMAKE_INSTALL_PREFIX=/usr
+            -DCMAKE_INSTALL_PREFIX=/usr \
+            -DALWAYS_ROOT=YES
       - name: Build
         run: |
           cmake --build build

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -69,14 +69,13 @@ jobs:
       - name: Build
         run: |
           cmake --build build
-      # New DIO test segfaults starting Apr 2026 see #2869
-      #- name: Test
-      #  run: |
-      #    cmake --build build --target test
-      #- name: Debug
-      #  if: failure()
-      #  run: |
-      #    cat build/Testing/Temporary/LastTest.log
+      - name: Test
+        run: |
+          cmake --build build --target test
+      - name: Debug
+        if: failure()
+        run: |
+          cat build/Testing/Temporary/LastTest.log
       - name: Install
         run: |
           DESTDIR=$PWD/build/install cmake --build build --target install
@@ -414,15 +413,13 @@ jobs:
       - name: Build
         run: |
           cmake --build build
-      # New DIO test segfaults starting Apr 2026 see #2869
-      #- name: Test
-      #  run: |
-      #    cmake --build build --target test
-      #- name: Debug Test Failure
-      #  if: failure()
-      #  run: |
-      #    build/test_dio -v2
-      #    cat build/Testing/Temporary/LastTest.log
+      - name: Test
+        run: |
+          cmake --build build --target test
+      - name: Debug Test Failure
+        if: failure()
+        run: |
+          cat build/Testing/Temporary/LastTest.log
       - name: Install
         run: |
           DESTDIR=$PWD/build/install cmake --build build --target install

--- a/.obs/debian.rules
+++ b/.obs/debian.rules
@@ -13,7 +13,3 @@
 # We don't download fonts as its a pkg dependency.
 override_dh_auto_configure:
 	dh_auto_configure -- -DFREECIV_DOWNLOAD_FONTS=OFF
-
-# We don't run tests in development per issue 2869
-override_dh_auto_test:
-	echo "Tests were skipped"

--- a/.obs/freeciv21.spec
+++ b/.obs/freeciv21.spec
@@ -10,7 +10,7 @@ Name:           freeciv21
 Version:        3.2
 Release:        %autorelease -p pre -s %{__scm_source_timestamp}
 Group:          Amusements/Games/Strategy/Turn Based
-Summary:        We develop and play Freeciv21. Develop Your Civilization from Humble Roots to a Global Empire!
+Summary:        Develop Your Civilization from Humble Roots to a Global Empire!
 Url:            https://github.com/longturn/freeciv21/
 Source:         %{name}-%{version}.tar.gz
 BuildRequires:  cmake

--- a/.obs/freeciv21.spec
+++ b/.obs/freeciv21.spec
@@ -104,6 +104,7 @@ Translation files for freeciv21.
 %cmake_install
 %find_lang %{name}-core
 %find_lang %{name}-nations
+%find_lang %{name}-ruledit
 %fdupes %{buildroot}/%{_datadir}/
 
 %files
@@ -125,6 +126,6 @@ Translation files for freeciv21.
 %{_datadir}/metainfo/net.longturn.%{name}.ruledit.metainfo.xml
 %{_mandir}/man6/%{name}-*
 
-%files lang -f %{name}-core.lang -f %{name}-nations.lang
+%files lang -f %{name}-core.lang -f %{name}-nations.lang -f %{name}-ruledit.lang
 
 %changelog

--- a/.obs/freeciv21.spec
+++ b/.obs/freeciv21.spec
@@ -95,8 +95,7 @@ Translation files for freeciv21.
 %setup -q -n %{name}-%{version}
 
 %build
-%cmake -DCMAKE_BUILD_TYPE=Debug \
-       -DFREECIV_DOWNLOAD_FONTS=OFF \
+%cmake -DFREECIV_DOWNLOAD_FONTS=OFF \
        -DCMAKE_INSTALL_DOCDIR=%{_docdir}/%{name}
 %cmake_build
 

--- a/.obs/freeciv21.spec
+++ b/.obs/freeciv21.spec
@@ -20,6 +20,8 @@ BuildRequires:  readline-devel
 BuildRequires:  zlib-devel
 BuildRequires:  fdupes
 BuildRequires:  pkgconfig(SDL2_mixer)
+# For checks
+BuildRequires:  desktop-file-utils
 
 %if 0%{?fedora}
 BuildRequires:  sqlite-devel
@@ -41,6 +43,8 @@ BuildRequires:  libvpx
 %ifarch x86_64
 BuildRequires:  libvpl >= 2.15
 %endif
+# For checks
+BuildRequires:  libappstream-glib
 Recommends:     freeciv21-lang
 Recommends:     alerque-libertinus-fonts
 %endif
@@ -55,6 +59,8 @@ BuildRequires:  lua53-devel
 BuildRequires:  python313-Sphinx
 Recommends:     freeciv21-lang
 Recommends:     libertinus-fonts
+# For checks
+BuildRequires:  appstream-glib
 %endif
 
 BuildRoot:      %{_tmppath}/%{name}-%{version}-build
@@ -105,6 +111,17 @@ Translation files for freeciv21.
 %find_lang %{name}-nations
 %find_lang %{name}-ruledit
 %fdupes %{buildroot}/%{_datadir}/
+
+%check
+appstream-util validate --nonet %{buildroot}/%{_datadir}/metainfo/net.longturn.%{name}.metainfo.xml
+appstream-util validate --nonet %{buildroot}/%{_datadir}/metainfo/net.longturn.%{name}.server.metainfo.xml
+appstream-util validate --nonet %{buildroot}/%{_datadir}/metainfo/net.longturn.%{name}.modpack.metainfo.xml
+appstream-util validate --nonet %{buildroot}/%{_datadir}/metainfo/net.longturn.%{name}.ruledit.metainfo.xml
+desktop-file-validate %{buildroot}/%{_datadir}/applications/net.longturn.%{name}.desktop
+desktop-file-validate %{buildroot}/%{_datadir}/applications/net.longturn.%{name}.modpack.desktop
+desktop-file-validate %{buildroot}/%{_datadir}/applications/net.longturn.%{name}.ruledit.desktop
+desktop-file-validate %{buildroot}/%{_datadir}/applications/net.longturn.%{name}.server.desktop
+%ctest
 
 %files
 %defattr(-,root,root,-)

--- a/common/networking/dataio_raw.h
+++ b/common/networking/dataio_raw.h
@@ -95,7 +95,7 @@ void dio_put(QByteArray &dout, T value, T = 0)
 template <class T, class = std::enable_if_t<!std::is_same_v<T, int>>>
 bool dio_get(QByteArrayView &din, int &dest, T = 0)
 {
-  T tmp;
+  T tmp = 0;
   auto ret = dio_get(din, tmp);
   dest = tmp;
   return ret;
@@ -117,7 +117,7 @@ void dio_put(QByteArray &dout, int value, T = 0)
  */
 inline bool dio_get(QByteArrayView &din, bool &dest)
 {
-  int tmp;
+  int tmp = 0;
   auto ret = dio_get<std::uint8_t>(din, tmp);
   dest = (tmp != 0);
   return ret;
@@ -139,7 +139,7 @@ inline void dio_put(QByteArray &dout, bool value)
 template <class T>
 bool dio_get(QByteArrayView &din, float &dest, int precision, T = 0)
 {
-  int ival;
+  int ival = 0;
   auto ret = dio_get<T>(din, ival);
   dest = static_cast<float>(ival) / precision;
   return ret;
@@ -165,7 +165,7 @@ template <class Enum, class T,
           class = std::enable_if_t<std::is_enum_v<Enum>>>
 bool dio_get(QByteArrayView &din, Enum &dest, T = 0)
 {
-  int ival;
+  int ival = 0;
   auto ret = dio_get<T>(din, ival);
   dest = static_cast<Enum>(ival);
   return ret;
@@ -241,7 +241,7 @@ bool dio_get(QByteArrayView &din, std::array<T, N> &dest, array_diff,
   // more than 256 values in an array-diff.
   for (int count = 0; count < 256; ++count) {
     // Get the index in the array.
-    std::uint8_t index;
+    std::uint8_t index = 0xff;
     fc_assert_ret_val_msg(dio_get(din, index), false,
                           "array-diff: incomplete data");
 

--- a/common/tests/dio.cpp
+++ b/common/tests/dio.cpp
@@ -204,5 +204,5 @@ void test_dio::string()
   QCOMPARE(din.size(), 0);
 }
 
-QTEST_MAIN(test_dio)
+QTEST_GUILESS_MAIN(test_dio)
 #include "dio.moc"

--- a/data/tests/rulesets.cpp.in
+++ b/data/tests/rulesets.cpp.in
@@ -43,6 +43,7 @@ void test_rulesets::load()
   QFETCH(bool, exists);
 
   QProcess p;
+  p.setProcessChannelMode(QProcess::ForwardedChannels);
   p.start(SERVER_PATH, {QStringLiteral("-r"), name});
   p.waitForStarted();
   p.write("quit\n");

--- a/server/tests/cli.cpp.in
+++ b/server/tests/cli.cpp.in
@@ -46,6 +46,7 @@ void test_cli::read()
   QFETCH(bool, success);
 
   QProcess p;
+  p.setProcessChannelMode(QProcess::ForwardedChannels);
   p.start(SERVER_PATH, {QStringLiteral("-r"), QStringLiteral(DATA_PATH) + file});
   p.waitForStarted();
   p.write("quit\n");

